File: android/jvm-tests/src/test/kotlin/AliasTest.kt
import hu.breaker.app.core.AliasLogic
import hu.breaker.app.core.Site
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * A fedőnév Kotlin-magja — a `desktop/test/alias.test.ts` tükre.
 *
 * Ugyanazok az esetek ugyanazokkal az elvárásokkal: ha a két mag elcsúszik, a
 * felhasználó ugyanazt az appot kapja két különböző viselkedéssel.
 */
class AliasTest {

    private fun site(domain: String, alias: String? = null) = Site(
        id = "site_1", domain = domain, hostnames = listOf(domain),
        addedAt = 0L, pauseUntil = null, pendingDeleteAt = null, alias = alias,
    )

    @Test
    fun `no alias means the domain is shown`() {
        assertEquals("youtube.com", AliasLogic.displayName(site("youtube.com")))
        assertEquals("youtube.com", AliasLogic.displayName(site("youtube.com", "")))
        assertEquals("youtube.com", AliasLogic.displayName(site("youtube.com", "   ")))
        assertFalse(AliasLogic.isAliased(site("youtube.com", "  ")))
    }

    @Test
    fun `an alias replaces the domain`() {
        val s = site("youtube.com", "A videós")
        assertEquals("A videós", AliasLogic.displayName(s))
        assertTrue(AliasLogic.isAliased(s))
    }

    @Test
    fun `control characters cannot hide inside an alias`() {
        val a = AliasLogic.normalize("A\u0000vide\u001Fós\u007F")
        assertEquals("A vide ós", a)
        assertFalse(a!!.any { it.code < 0x20 || it.code in 0x7f..0x9f })
    }

    @Test
    fun `whitespace is collapsed and trimmed`() {
        assertEquals("A videós", AliasLogic.normalize("  A    videós  "))
        assertEquals("A videós", AliasLogic.normalize("\n\tA videós\n"))
    }

    @Test
    fun `a very long alias is cut, and never left with a trailing space`() {
        val long = "x".repeat(AliasLogic.MAX_ALIAS_LENGTH + 20)
        assertEquals(AliasLogic.MAX_ALIAS_LENGTH, AliasLogic.normalize(long)!!.length)
        // A vágás szóköz közepére eshet; a maradék végén ne maradjon lógó szóköz.
        val spaced = "a".repeat(AliasLogic.MAX_ALIAS_LENGTH - 1) + " bbbb"
        assertEquals("a".repeat(AliasLogic.MAX_ALIAS_LENGTH - 1), AliasLogic.normalize(spaced))
    }

    @Test
    fun `nonsense input is treated as no alias`() {
        assertNull(AliasLogic.normalize(null))
        assertNull(AliasLogic.normalize(" "))
        assertNull(AliasLogic.normalize(""))
    }

    @Test
    fun `a reveal shows the real domain, but only while it lasts`() {
        val s = site("youtube.com", "A videós")
        val now = 1_000_000L
        val until = now + AliasLogic.REVEAL_MS
        assertEquals("youtube.com", AliasLogic.displayNameNow(s, now, until))
        assertEquals("youtube.com", AliasLogic.displayNameNow(s, until - 1, until))
        assertEquals("A videós", AliasLogic.displayNameNow(s, until, until))
        assertEquals("A videós", AliasLogic.displayNameNow(s, now + 60_000, until))
        assertEquals("A videós", AliasLogic.displayNameNow(s, now, null))
    }

    @Test
    fun `a reveal on a site with no alias changes nothing`() {
        assertEquals("reddit.com", AliasLogic.displayNameNow(site("reddit.com"), 0L, 9_999_999L))
    }

    @Test
    fun `a hidden list masks the domain in statistics, but an alias wins`() {
        assertEquals("1. rejtett oldal", AliasLogic.maskedLabel(site("youtube.com"), 0))
        assertEquals("3. rejtett oldal", AliasLogic.maskedLabel(site("reddit.com"), 2))
        assertEquals("A videós", AliasLogic.maskedLabel(site("youtube.com", "A videós"), 0))
    }
}

File: desktop/test/alias.test.ts
import { test } from 'node:test';
import * as assert from 'node:assert/strict';
import {
  displayName, displayNameNow, isAliased, normalizeAlias, MAX_ALIAS_LENGTH, REVEAL_MS,
} from '../src/shared/alias';

test('no alias means the domain is shown', () => {
  assert.equal(displayName({ domain: 'youtube.com' }), 'youtube.com');
  assert.equal(displayName({ domain: 'youtube.com', alias: '' }), 'youtube.com');
  assert.equal(displayName({ domain: 'youtube.com', alias: '   ' }), 'youtube.com');
  assert.equal(isAliased({ domain: 'youtube.com', alias: '  ' }), false);
});

test('an alias replaces the domain', () => {
  const site = { domain: 'youtube.com', alias: 'A videós' };
  assert.equal(displayName(site), 'A videós');
  assert.equal(isAliased(site), true);
});

test('control characters cannot hide inside an alias', () => {
  // Láthatatlanok maradnának a soron, de a hosszkorlátba beleszámítanának, és a
  // mentett állapotba is bekerülnének.
  const a = normalizeAlias('A\u0000vide\u001fós\u007f');
  assert.equal(a, 'A vide ós');
  assert.ok(!/[\u0000-\u001f\u007f-\u009f]/.test(a!));
});

test('whitespace is collapsed and trimmed', () => {
  assert.equal(normalizeAlias('  A    videós  '), 'A videós');
  assert.equal(normalizeAlias('\n\tA videós\n'), 'A videós');
});

test('a very long alias is cut, and never left with a trailing space', () => {
  const long = 'x'.repeat(MAX_ALIAS_LENGTH + 20);
  assert.equal(normalizeAlias(long)!.length, MAX_ALIAS_LENGTH);
  // A vágás szóköz közepére eshet; a maradék végén ne maradjon lógó szóköz.
  const spaced = `${'a'.repeat(MAX_ALIAS_LENGTH - 1)} bbbb`;
  assert.equal(normalizeAlias(spaced), 'a'.repeat(MAX_ALIAS_LENGTH - 1));
});

test('nonsense input is treated as no alias', () => {
  assert.equal(normalizeAlias(undefined), undefined);
  assert.equal(normalizeAlias(null), undefined);
  assert.equal(normalizeAlias(42 as unknown as string), undefined);
  assert.equal(normalizeAlias(' '), undefined);
});

test('a reveal shows the real domain, but only while it lasts', () => {
  const site = { domain: 'youtube.com', alias: 'A videós' };
  const now = 1_000_000;
  assert.equal(displayNameNow(site, now, now + REVEAL_MS), 'youtube.com', 'felfedés alatt a cím');
  assert.equal(displayNameNow(site, now + REVEAL_MS - 1, now + REVEAL_MS), 'youtube.com');
  assert.equal(displayNameNow(site, now + REVEAL_MS, now + REVEAL_MS), 'A videós', 'lejárva újra fedőnév');
  assert.equal(displayNameNow(site, now + 60_000, now + REVEAL_MS), 'A videós');
  assert.equal(displayNameNow(site, now, undefined), 'A videós', 'felfedés nélkül sosem a cím');
});

test('a reveal on a site with no alias changes nothing', () => {
  assert.equal(displayNameNow({ domain: 'reddit.com' }, 0, 9_999_999), 'reddit.com');
});
